### PR TITLE
Removed intermediate wire between instances and output operations

### PR DESCRIPTION
**Pull request by jenra-uwu**
Before, the MLIR code below produced the following Verilog code:
```
hw.module.extern @B() -> (%B_output : i32)

hw.module @A() -> (%A_output : i32) {
    %0 = hw.instance "b1" @B() : () -> (i32)
    hw.output %0 : i32
}
```
```
// external module B

module A(
  output [31:0] A_output);

  wire b1_b_output

  B b1 (        // ../test.mlir:4:10
    .B_output (b1_b_output)
  );

  assign A_output = b1_b_output;
endmodule
```

With this pull request, the above MLIR code becomes:
```
// external module B

module A(
  output [31:0] A_output);

  B b1 (        // ../test.mlir:4:10
    .B_output (A_output)
  );
endmodule
```

The intermediate wire is removed and the output of the instance is directly assigned to the output of the module. This does not apply to instances where the output of an instance is used in multiple places:
```
hw.module.extern @B() -> (%B_output : i32)
hw.module.extern @C(%c_in : i32) -> ()

hw.module @A() -> (%A_output : i32) {
    %0 = hw.instance "b1" @B() : () -> (i32)
    hw.instance "c1" @C(%0) : (i32) -> ()
    hw.output %0 : i32
}
```
```
// external module B

// external module C

module A(
  output [31:0] A_output);

  wire [31:0] b1_B_output;      // ../test.mlir:5:10

  B b1 (        // ../test.mlir:5:10
    .B_output (b1_B_output)
  );
  C c1 (        // ../test.mlir:6:5
    .c_in (b1_B_output) // ../test.mlir:5:10
  );
  assign A_output = b1_B_output;        // ../test.mlir:5:10, :7:5
endmodule
```

### CODE REVIEW

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 1]
```diff
@@ -1844,13 +1845,21 @@ LogicalResult StmtEmitter::visitStmt(OutputOp op) {
...
+    }
+
     ops.clear();
     ops.insert(op);
     indent();
     if (isZeroBitType(port.type))
       os << "// Zero width: ";
     os << "assign " << emitter.outputNames[port.argNum] << " = ";
-    emitExpression(op.getOperand(operandIndex), ops);
+    emitExpression(operand, ops);
```
Comment (fabianschuiki) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:1862]: Does your change still work if the port is of zero width? That could be a neat additional test case :+1:

**Review by fabianschuiki — COMMENTED** [review]
Nice! Thanks for tackling this :+1:. One comment inline regarding zero width ports.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp (hunk 1 above)
Comment (jenra-uwu) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:1862]: That's something I have not tested; I'll try it out now.

Comment (jenra-uwu) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:1862]: I've just tested it, and it looks like the wires are correctly removed, but I don't think the generated Verilog is valid since the right parenthesis is in a comment.
```
hw.module.extern @B() -> (%B_output : i0)

hw.module @A() -> (%A_output : i0) {
    %0 = hw.instance "b1" @B() : () -> (i0)
    hw.output %0 : i0
}
```
```
// external module B

module A(
  // output /*Zero Width*/ A_output
   );

  B b1 (        // ../test.mlir:5:10
  //.B_output (A_output));
endmodule
```

Comment (fabianschuiki) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:1862]: Cool! This might be an issue with the existing module emission. Might be a worthwhile follow-up PR on ExportVerilog to specifically fix this issue (unrelated to your nice fix here). I'd still recommend adding your test case for zero-width ports; you should be able to just do something like `// CHECK: //.B_output (A_output)` to not have the Verilog emission issue affect you :wink:

**Review by fabianschuiki — APPROVED** [review]
Looks good to me. Let's see what the CI build thinks of this (should be running now).

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 2]
```diff
@@ -2399,10 +2408,28 @@ LogicalResult StmtEmitter::visitStmt(InstanceOp op) {
 
     // Emit the value as an expression.
     ops.clear();
-    // output ports were lowered to wire
-    if (elt.isOutput())
+    // output ports that are not connected to single use output ports were lowered to wire
+    OutputOp output = dyn_cast_or_null<OutputOp>(portVal.getUses().begin()->getOwner());
```
Comment (darthscsi) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:2412]: I don't think you always have even one use.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 3]
```diff
@@ -2399,10 +2408,28 @@ LogicalResult StmtEmitter::visitStmt(InstanceOp op) {
...
+    if (elt.isOutput() && portVal.hasOneUse() && output) {
+      auto module = output->getParentOfType<HWModuleOp>();
+      SmallVector<ModulePortInfo> portInfo = getModulePortInfo(module);
+
+      for (size_t portNum = 0, operandNum = 0; portNum < portInfo.size(); ++portNum) {
```
Comment (darthscsi) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:2417]: I think you can use getModuleResultNameAttr here.

**Review by darthscsi — APPROVED** [review]

File: lib/Translation/ExportVerilog/ExportVerilog.cpp (hunk 3 above)
Comment (jenra-uwu) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:2417]: How do you get the operand number for the second argument of `getModuleResultNameAttr`?

File: lib/Translation/ExportVerilog/ExportVerilog.cpp (hunk 2 above)
Comment (jenra-uwu) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:2412]: I think the `portVal` has at least one use, which is assigning it to the output port of the parent module.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 4]
```diff
@@ -20,6 +20,7 @@
 #include "circt/Dialect/SV/SVVisitors.h"
 #include "circt/Support/LLVM.h"
 #include "circt/Support/LoweringOptions.h"
+#include "mlir/IR/AsmState.h"
```
Comment (lattner) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:23]: I don't think AsmState.h is needed?

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 5]
```diff
@@ -1898,13 +1899,21 @@ LogicalResult StmtEmitter::visitStmt(OutputOp op) {
   for (ModulePortInfo port : parent.getPorts()) {
     if (!port.isOutput())
       continue;
+
+    auto operand = op.getOperand(operandIndex);
+    if (operand.hasOneUse() && dyn_cast_or_null<InstanceOp>(operand.getDefiningOp())) {
+      ++operandIndex;
+      ++numStatementsEmitted;
```
Comment (lattner) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:1906]: Please don't increment numStatementsEmitted.  It doesn't really matter here, but this should track the number of verilog statements emitted and this isn't doing it.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp (hunk 2 above)
Comment (lattner) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:2412]: I don't think this is correct for non-output ports, which can have zero users.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp (hunk 3 above)
Comment (lattner) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:2417]: I think that portVal is guaranteed to be a `OpResult`, which has a `getResultNumber()` method.  Does that work?

Comment (lattner) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:2417]: I agree with darth that computing the number from the IR would be better.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 6]
```diff
@@ -2399,10 +2408,28 @@ LogicalResult StmtEmitter::visitStmt(InstanceOp op) {
...
+
+        if (output->getOperand(operandNum) == portVal) {
+          os << portInfo[portNum].getName();
+          break;
+        }
+
+        ++operandNum;
+      }
+
+    } else if (elt.isOutput()) {
```
Comment (lattner) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:2429]: I'd recommend pulling the "isOutput" higher since it encompases both checks.  You could even invert this logic a bit to be more readable, something like:

```
if (!elt.isOutput()) {
  emitExpression(portVal, ops);
} else if (<not the inline case>) {
  emitExpression(getWireForValue(portVal), ops);
} else {
  ...
  os << the name
}
```

**Review by lattner — COMMENTED** [review]
Awesome, thank you for working on this!  I'm very excited to see this get improved!

File: lib/Translation/ExportVerilog/ExportVerilog.cpp (hunk 4 above)
Comment (jenra-uwu) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:23]: Ah, must have automatically been imported by Neovim for something I removed and then I forgot about it, sorry!

File: lib/Translation/ExportVerilog/ExportVerilog.cpp (hunk 3 above)
Comment (jenra-uwu) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp]: Oh, apparently it starts at 1 instead of at 0. :/

Comment (jenra-uwu) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp]: Okay it's still not working...

Comment (jenra-uwu) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp]: Okay I fixed it!

**Comment by jenra-uwu** [comment]
@lattner Could you check my pull request again? I've gone through everything you've suggested.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 7]
```diff
@@ -2399,10 +2406,20 @@ LogicalResult StmtEmitter::visitStmt(InstanceOp op) {
 
     // Emit the value as an expression.
     ops.clear();
-    // output ports were lowered to wire
-    if (elt.isOutput())
+
+    // output ports that are not connected to single use output ports were lowered to wire
```
Comment (lattner) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:2410]: Super nit, but please start sentence comments with a capital letter and end with a period.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 8]
```diff
@@ -2399,10 +2406,20 @@ LogicalResult StmtEmitter::visitStmt(InstanceOp op) {
...
+    OutputOp output;
+    if (!elt.isOutput())
+      emitExpression(portVal, ops);
+    else if (portVal.hasOneUse() && (output = dyn_cast_or_null<OutputOp>(portVal.getUses().begin()->getOwner()))) {
```
Comment (lattner) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:2414]: Plz fit to 80 columns (run clang format).  Also, please put braces around the previous "if" to be consistent with the else.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 9]
```diff
@@ -2399,10 +2406,20 @@ LogicalResult StmtEmitter::visitStmt(InstanceOp op) {
...
+      auto module = output->getParentOfType<HWModuleOp>();
+      SmallVector<ModulePortInfo> portInfo = getModulePortInfo(module);
```
Comment (lattner) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:2416]: You can delete portInfo now.

**Review by lattner — APPROVED** [review]
Looks really great!  Feel free to commit after tidying up the nit picky things I mention above, thank you again!

**Comment by lattner** [comment]
Thanks, feel free to land whenever you're ready.  If you need commit access, please request through the LLVM Developer Policy process
